pallets/pools/src/tranche.rs

```diff
@@ -233,64 +224,377 @@ where
+	Tranches<Balance, Rate, Weight, CurrencyId, TrancheId, PoolId>
 where
 	CurrencyId: Copy,
 	Balance: Zero + Copy + BaseArithmetic + FixedPointOperand + Unsigned + From<u64>,
 	Weight: Copy + From<u128>,
 	Rate: One + Copy + FixedPointNumber<Inner = Balance>,
+	TrancheId: Clone + From<[u8; 16]> + sp_std::cmp::PartialEq,
+	PoolId: Copy + Encode,
 {
-	pub fn from_input<
-		PoolId: Copy,
-		TrancheId: TryFrom<usize>,
-		TrancheToken: TrancheTokenT<PoolId, TrancheId, CurrencyId>,
-	>(
-		pool_id: PoolId,
+	pub fn from_input<TrancheToken>(
+		pool: PoolId,
 		tranche_inputs: Vec<TrancheInput<Rate>>,
 		now: Moment,
-	) -> Result<Self, DispatchError> {
+	) -> Result<Self, DispatchError>
+	where
+		TrancheToken: TrancheTokenT<PoolId, TrancheId, CurrencyId>,
+	{
 		let mut tranches = Vec::with_capacity(tranche_inputs.len());
-
-		for (id, (tranche_type, seniority)) in tranche_inputs.into_iter().enumerate() {
+		let mut ids = Vec::with_capacity(tranche_inputs.len());
+		let mut salt = (0, pool);
+
+		for (index, (tranche_type, seniority)) in tranche_inputs.into_iter().enumerate() {
+			// Generate ids after the following schema:
+			// * salt: The salt is a counter in our case that will always go
+			//         up, even if we remove tranches.
+			// * pool-id: The pool id is ensured to be unique on-chain
+			//
+			// -> tranche id = Twox128::hash(salt)
+			let hash_input = salt.encode();
+			let id = Twox128::hash(hash_input.as_slice());
+
+			ids.push(id.into());
 			tranches.push(Tranche {
 				tranche_type,
 				// seniority increases as index since the order is from junior to senior
 				seniority: seniority
-					.unwrap_or(id.try_into().map_err(|_| ArithmeticError::Overflow)?),
-				currency: TrancheToken::tranche_token(
-					pool_id,
-					id.try_into().map_err(|_| ArithmeticError::Overflow)?,
-				),
-
+					.unwrap_or(index.try_into().map_err(|_| ArithmeticError::Overflow)?),
+				currency: TrancheToken::tranche_token(pool, id.into()),
 				outstanding_invest_orders: Zero::zero(),
 				outstanding_redeem_orders: Zero::zero(),
-
 				debt: Zero::zero(),
 				reserve: Zero::zero(),
 				ratio: Perquintill::zero(),
 				last_updated_interest: now,
 				_phantom: Default::default(),
-			})
+			});
+
+			salt = (
+				(index.checked_add(1).ok_or(ArithmeticError::Overflow)?)
+					.try_into()
+					.map_err(|_| ArithmeticError::Overflow)?,
+				pool,
+			);
 		}
 
-		Ok(Self { tranches })
+		Ok(Self {
+			tranches,
+			ids,
+			salt,
+		})
 	}
 
-	pub fn new(tranches: Vec<Tranche<Balance, Rate, Weight, CurrencyId>>) -> Self {
-		Self { tranches }
+	pub fn new<TrancheToken>(
+		pool: PoolId,
+		tranches: Vec<Tranche<Balance, Rate, Weight, CurrencyId>>,
+	) -> Result<Self, DispatchError>
+	where
+		TrancheToken: TrancheTokenT<PoolId, TrancheId, CurrencyId>,
+	{
+		let mut ids = Vec::with_capacity(tranches.len());
+		let mut salt = (0, pool);
+
+		for (index, _tranche) in tranches.iter().enumerate() {
+			// Generate ids after the following schema:
```

Comment:
It might be nice to move this into a separate method so it can be re-used between `from_input()` and `new()`?

Reply:
Yeah, I know. Will do. Haha.

Reply:
Updated.